pkg/gardenlet/controller/tokenrequestor/workloadidentity/reconciler.go

```diff
@@ -0,0 +1,157 @@
+	"k8s.io/utils/clock"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	securityv1alpha1 "github.com/gardener/gardener/pkg/apis/security/v1alpha1"
+	securityv1alpha1constants "github.com/gardener/gardener/pkg/apis/security/v1alpha1/constants"
+	securityclientset "github.com/gardener/gardener/pkg/client/security/clientset/versioned"
+	"github.com/gardener/gardener/pkg/controllerutils"
+)
+
+const (
+	tokenRenewTimestamp   = securityv1alpha1constants.WorkloadIdentityPrefix + "token-renew-timestamp"
+	maxExpirationDuration = 24 * time.Hour
+	expirationDuration    = 6 * time.Hour // short enough to be secure and long enough to be resilient to disruptions
+)
+
+// Reconciler requests and refreshes tokens via the TokenRequest API.
+type Reconciler struct {
+	SeedClient           client.Client
+	GardenClient         client.Client
+	GardenSecurityClient securityclientset.Interface
+	ConcurrentSyncs      int
+	Clock                clock.Clock
+	JitterFunc           func(time.Duration, float64) time.Duration
+}
+
+// Reconcile requests and populates tokens.
+func (r *Reconciler) Reconcile(reconcileCtx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := logf.FromContext(reconcileCtx)
+
+	ctx, cancel := controllerutils.GetMainReconciliationContext(reconcileCtx, controllerutils.DefaultReconciliationTimeout)
+	defer cancel()
+
+	secret := &corev1.Secret{}
+	if err := r.SeedClient.Get(ctx, req.NamespacedName, secret); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.V(1).Info("Object is gone, stop reconciling")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("error retrieving object from store: %w", err)
+	}
+
+	if !r.isRelevantSecret(secret) {
+		return reconcile.Result{}, nil
+	}
+
+	mustRequeue, requeueAfter, err := r.requeue(secret)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if mustRequeue {
+		log.Info("No need to generate new token, renewal is scheduled", "after", requeueAfter)
+		return reconcile.Result{Requeue: true, RequeueAfter: requeueAfter}, nil
+	}
+
+	log.Info("Requesting new token")
+
+	var contextObject *securityv1alpha1.ContextObject
+	if v, ok := secret.Annotations[securityv1alpha1constants.AnnotationWorkloadIdentityContextObject]; ok {
+		contextObject = &securityv1alpha1.ContextObject{}
+		if err := json.Unmarshal([]byte(v), contextObject); err != nil {
+			return reconcile.Result{}, fmt.Errorf("cannot parse context object: %w", err)
+		}
+	}
+
+	workloadIdentity := &securityv1alpha1.WorkloadIdentity{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secret.Annotations[securityv1alpha1constants.AnnotationWorkloadIdentityName],
+			Namespace: secret.Annotations[securityv1alpha1constants.AnnotationWorkloadIdentityNamespace],
+		},
+	}
+
+	if err := r.GardenClient.Get(ctx, client.ObjectKeyFromObject(workloadIdentity), workloadIdentity); err != nil {
```

Comment:
Is it possible to use controller runtime Subresource client?
```go
	r.GardenClient.SubResource("token").Create(ctx, workloadIdentity,
		&securityv1alpha1.TokenRequest{
			Spec: securityv1alpha1.TokenRequestSpec{
				ContextObject:     contextObject,
				ExpirationSeconds: ptr.To((int64(expirationDuration / time.Second))),
			},
		},
	)
```

Reply:
I left a comment, but right now it is not easily testable with the fake client implementation. Similar to https://github.com/gardener/gardener/blob/e597a52eedf971265c55cab86dec6c1599d31984/pkg/controller/tokenrequestor/reconciler.go#L188